<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.313">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Coriell Bioinformatics Notes - ATAC-seq Analysis</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./references.html" rel="next">
<link href="./differential-methylation-analysis.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>


</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
    <div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title">ATAC-seq Analysis</h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Coriell Bioinformatics Notes</a> 
    </div>
      </div>
      <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
      </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">Bioinformatics Notes</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./data-best-practices.html" class="sidebar-item-text sidebar-link">‘Best’ practices for data projects</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./linux.html" class="sidebar-item-text sidebar-link">Command line basics</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./r-basics.html" class="sidebar-item-text sidebar-link">R programming basics</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./data-viz.html" class="sidebar-item-text sidebar-link">Effective data visualizations</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./presenting-science.html" class="sidebar-item-text sidebar-link">Scientific presentations</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./summarized-experiments.html" class="sidebar-item-text sidebar-link">Practical introduction to SummarizedExperiments</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./differential-expression-analysis.html" class="sidebar-item-text sidebar-link">Differential Expression Analysis</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./differential-methylation-analysis.html" class="sidebar-item-text sidebar-link">Differential Methylation Analysis</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./atacseq-analysis.html" class="sidebar-item-text sidebar-link active">ATAC-seq Analysis</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">References</a>
  </div>
</li>
    </ul>
    </div>
</nav>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#filter-raw-fastq-files" id="toc-filter-raw-fastq-files" class="nav-link active" data-scroll-target="#filter-raw-fastq-files">Filter raw fastq files</a></li>
  <li><a href="#perform-alignment-with-bowtie2" id="toc-perform-alignment-with-bowtie2" class="nav-link" data-scroll-target="#perform-alignment-with-bowtie2">Perform alignment with Bowtie2</a></li>
  <li><a href="#call-peaks-with-genrich" id="toc-call-peaks-with-genrich" class="nav-link" data-scroll-target="#call-peaks-with-genrich">Call peaks with <code>Genrich</code></a>
  <ul class="collapse">
  <li><a href="#exclusion-lists" id="toc-exclusion-lists" class="nav-link" data-scroll-target="#exclusion-lists">Exclusion lists</a></li>
  </ul></li>
  <li><a href="#determine-consensus-peaks-for-replicate-samples" id="toc-determine-consensus-peaks-for-replicate-samples" class="nav-link" data-scroll-target="#determine-consensus-peaks-for-replicate-samples">Determine consensus peaks for replicate samples</a></li>
  <li><a href="#create-raw-signal-tracks" id="toc-create-raw-signal-tracks" class="nav-link" data-scroll-target="#create-raw-signal-tracks">Create raw signal tracks</a>
  <ul class="collapse">
  <li><a href="#chrom-sizes-files" id="toc-chrom-sizes-files" class="nav-link" data-scroll-target="#chrom-sizes-files">Chrom sizes files</a></li>
  </ul></li>
  <li><a href="#atacseqqc" id="toc-atacseqqc" class="nav-link" data-scroll-target="#atacseqqc">ATACSeqQC</a></li>
  <li><a href="#differential-abundance-using-csaw-and-edger" id="toc-differential-abundance-using-csaw-and-edger" class="nav-link" data-scroll-target="#differential-abundance-using-csaw-and-edger">Differential abundance using <code>csaw</code> and <code>edgeR</code></a></li>
  <li><a href="#create-tss-and-region-plots-with-deeptools" id="toc-create-tss-and-region-plots-with-deeptools" class="nav-link" data-scroll-target="#create-tss-and-region-plots-with-deeptools">Create TSS and region plots with <code>deeptools</code></a>
  <ul class="collapse">
  <li><a href="#getting-tss-coding-region-ranges" id="toc-getting-tss-coding-region-ranges" class="nav-link" data-scroll-target="#getting-tss-coding-region-ranges">Getting TSS coding region ranges</a></li>
  </ul></li>
  <li><a href="#motif-analysis-with-meme" id="toc-motif-analysis-with-meme" class="nav-link" data-scroll-target="#motif-analysis-with-meme">Motif analysis with <code>MEME</code></a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title d-none d-lg-block">ATAC-seq Analysis</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<p><strong>Work in progress</strong></p>
<section id="filter-raw-fastq-files" class="level2">
<h2 class="anchored" data-anchor-id="filter-raw-fastq-files">Filter raw fastq files</h2>
<p>Use <a href="https://github.com/OpenGene/fastp">fastp</a> to perform quality trimming on raw fastq files. Turn on adapter detection for paired end reads.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co">#!/usr/bin/env bash</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Run fastp on the raw fastq files</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="co"># ----------------------------------------------------------------------------</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="bu">set</span> <span class="at">-Eeou</span> pipefail</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="va">FQ</span><span class="op">=</span>/path/to/raw-fastq/00_fastq</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="va">SAMPLES</span><span class="op">=</span>/path/to/sample-names.txt   </span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="va">OUT</span><span class="op">=</span>/path/to/01_fastp</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="va">THREADS</span><span class="op">=</span>12</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> <span class="va">$OUT</span></span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> SAMPLE <span class="kw">in</span> <span class="va">$(</span><span class="fu">cat</span> <span class="va">$SAMPLES)</span></span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a><span class="cf">do</span></span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a>    <span class="ex">fastp</span> <span class="at">-i</span> <span class="va">$FQ</span>/<span class="va">${SAMPLE}</span>_R1.fastq.gz <span class="dt">\</span></span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a>          <span class="at">-I</span> <span class="va">$FQ</span>/<span class="va">${SAMPLE}</span>_R2.fastq.gz <span class="dt">\</span></span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a>          <span class="at">-o</span> <span class="va">$OUT</span>/<span class="va">${SAMPLE}</span>.trimmed.1.fq.gz <span class="dt">\</span></span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a>          <span class="at">-O</span> <span class="va">$OUT</span>/<span class="va">${SAMPLE}</span>.trimmed.2.fq.gz <span class="dt">\</span></span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true" tabindex="-1"></a>          <span class="at">-h</span> <span class="va">$OUT</span>/<span class="va">${SAMPLE}</span>.fastp.html <span class="dt">\</span></span>
<span id="cb1-22"><a href="#cb1-22" aria-hidden="true" tabindex="-1"></a>          <span class="at">-j</span> <span class="va">$OUT</span>/<span class="va">${SAMPLE}</span>.fastp.json <span class="dt">\</span></span>
<span id="cb1-23"><a href="#cb1-23" aria-hidden="true" tabindex="-1"></a>            <span class="at">--detect_adapter_for_pe</span> <span class="dt">\</span></span>
<span id="cb1-24"><a href="#cb1-24" aria-hidden="true" tabindex="-1"></a>          <span class="at">-w</span> <span class="va">$THREADS</span></span>
<span id="cb1-25"><a href="#cb1-25" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><code>sample-names.txt</code> is a plain text file listing the basenames for all samples. For example, if you have “sample1_R1.fq.gz”, “sample2_R1.fq.gz”, “sample3_R1.fq.gz” then <code>sample-names.txt</code> would be:</p>
<pre><code>sample1
sample2
sample3</code></pre>
<p>This file gets used throughout.</p>
</section>
<section id="perform-alignment-with-bowtie2" class="level2">
<h2 class="anchored" data-anchor-id="perform-alignment-with-bowtie2">Perform alignment with Bowtie2</h2>
<p>Use <code>--very-sensitive</code> mode and set max insertion size to 1,000. The output is piped to <code>samtools fixmate</code> for adding mate tag information and then to <code>samtools sort -n</code> to create name sorted BAM files.</p>
<p>It’s assumed that you have a bowtie2 index generated for your species of interest. The number of jobs, threads, and memory allowed per thread will all be machine dependent.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co">#!/usr/bin/env bash</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Align trimmed reads with bowtie2, fixmate tags, and output name sorted BAMs</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="co"># ----------------------------------------------------------------------------</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="bu">set</span> <span class="at">-Eeou</span> pipefail</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a><span class="va">FQ</span><span class="op">=</span>/path/to/01_fastp</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a><span class="va">SAMPLES</span><span class="op">=</span>/path/to/sample-names.txt   </span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a><span class="va">OUT</span><span class="op">=</span>/path/to/02_align</span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a><span class="va">IDX</span><span class="op">=</span>/path/to/bt2_idx</span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a><span class="va">INSERT</span><span class="op">=</span>1000</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a><span class="va">JOBS</span><span class="op">=</span>6</span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a><span class="va">THREADS</span><span class="op">=</span>4</span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> <span class="va">$OUT</span></span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a><span class="ex">parallel</span> <span class="at">--jobs</span> <span class="va">$JOBS</span> <span class="dt">\</span></span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a>  <span class="st">"bowtie2 -x </span><span class="va">$IDX</span><span class="st"> </span><span class="dt">\</span></span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a><span class="st">    -1 </span><span class="va">$FQ</span><span class="st">/{}.trimmed.1.fq.gz </span><span class="dt">\</span></span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true" tabindex="-1"></a><span class="st">    -2 </span><span class="va">$FQ</span><span class="st">/{}.trimmed.2.fq.gz </span><span class="dt">\</span></span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true" tabindex="-1"></a><span class="st">    --very-sensitive </span><span class="dt">\</span></span>
<span id="cb3-23"><a href="#cb3-23" aria-hidden="true" tabindex="-1"></a><span class="st">    --threads </span><span class="va">$THREADS</span><span class="st"> </span><span class="dt">\</span></span>
<span id="cb3-24"><a href="#cb3-24" aria-hidden="true" tabindex="-1"></a><span class="st">    --maxins </span><span class="va">$INSERT</span><span class="st"> | </span><span class="dt">\</span></span>
<span id="cb3-25"><a href="#cb3-25" aria-hidden="true" tabindex="-1"></a><span class="st">    samtools fixmate -m -@</span><span class="va">$THREADS</span><span class="st"> - - | </span><span class="dt">\</span></span>
<span id="cb3-26"><a href="#cb3-26" aria-hidden="true" tabindex="-1"></a><span class="st">    samtools sort -n -@</span><span class="va">$THREADS</span><span class="st"> -m4G -o </span><span class="va">$OUT</span><span class="st">/{}.bam - "</span> :::: <span class="va">$SAMPLES</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="call-peaks-with-genrich" class="level2">
<h2 class="anchored" data-anchor-id="call-peaks-with-genrich">Call peaks with <code>Genrich</code></h2>
<p><a href="https://github.com/jsh58/Genrich">Genrich</a> is a fast peak caller designed to work with single samples or replicates. It has a dedicated option for peak calling ATAC-seq samples.</p>
<p>The options used below:</p>
<ul>
<li>Output pileups over called regions in a pseudo bedfile (-k)</li>
<li>Output the log file (-f) which can be used to recall peaks with different -p and -a values</li>
<li>Exclude regions in blacklist (-E)</li>
<li>Exclude reads mapping to chrM (-e)</li>
<li>Call in ATAC-seq mode (-j)</li>
<li>Remove PCR duplicates (-r)</li>
<li>Specify p-value threshold for peak calling (-p)</li>
<li>Skip name sorting check (-S)</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co">#!/usr/bin/env bash</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Call ATAC-seq peaks using Genrich on the name sorted BAMs</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="co"># ----------------------------------------------------------------------------</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a><span class="bu">set</span> <span class="at">-Eeou</span> pipefail</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a><span class="va">SAMPLES</span><span class="op">=</span>/path/to/sample-names.txt   </span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a><span class="va">BAM</span><span class="op">=</span>/path/to/02_align</span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a><span class="va">OUT</span><span class="op">=</span>/path/to/03_callpeak</span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a><span class="va">EXCLUDE</span><span class="op">=</span>/path/to/excluderanges.bed</span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a><span class="va">PVAL</span><span class="op">=</span>0.05</span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a><span class="va">JOBS</span><span class="op">=</span>6</span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a><span class="ex">parallel</span> <span class="at">--jobs</span> <span class="va">$JOBS</span> <span class="dt">\</span></span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Genrich -t </span><span class="va">$BAM</span><span class="st">/{}.bam </span><span class="dt">\</span></span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a><span class="st">        -o </span><span class="va">$OUT</span><span class="st">/{}.narrowPeak </span><span class="dt">\</span></span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a><span class="st">        -k </span><span class="va">$OUT</span><span class="st">/{}.pileup.txt </span><span class="dt">\</span></span>
<span id="cb4-19"><a href="#cb4-19" aria-hidden="true" tabindex="-1"></a><span class="st">        -f </span><span class="va">$OUT</span><span class="st">/{}.log </span><span class="dt">\</span></span>
<span id="cb4-20"><a href="#cb4-20" aria-hidden="true" tabindex="-1"></a><span class="st">        -E </span><span class="va">$EXCLUDE</span><span class="st"> </span><span class="dt">\</span></span>
<span id="cb4-21"><a href="#cb4-21" aria-hidden="true" tabindex="-1"></a><span class="st">        -p </span><span class="va">$PVAL</span><span class="st"> -j -r -e chrM -S "</span> :::: <span class="va">$SAMPLES</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="exclusion-lists" class="level3">
<h3 class="anchored" data-anchor-id="exclusion-lists">Exclusion lists</h3>
<p>The “excluderanges.bed” file used above contains blacklisted regions of high-mappability that can be excluded from peak calling. This BED file can be created using the <a href="https://www.bioconductor.org/packages/release/data/annotation/vignettes/excluderanges/inst/doc/excluderanges.html">excluderanges</a> R package. For example, to generate the excluded regions for mm39:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(GenomicRanges))</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(AnnotationHub))</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>ah <span class="ot">&lt;-</span> <span class="fu">AnnotationHub</span>()</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>query_data <span class="ot">&lt;-</span> <span class="fu">subset</span>(ah, preparerclass <span class="sc">==</span> <span class="st">"excluderanges"</span>)</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>mm39_exclude_gr <span class="ot">&lt;-</span> query_data[[<span class="st">"AH107321"</span>]]</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>rtracklayer<span class="sc">::</span><span class="fu">export.bed</span>(mm39_exclude_gr, <span class="st">"mm39-excluderanges.bed"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="determine-consensus-peaks-for-replicate-samples" class="level2">
<h2 class="anchored" data-anchor-id="determine-consensus-peaks-for-replicate-samples">Determine consensus peaks for replicate samples</h2>
<p>After peak calling finishes, I create consensus peak calls for replicate samples. The consensus peak calls can be used in downstream differential accessibility analysis or for visualization. Creating consensus peak calls can be done in R by requiring all or some peaks to be called across replicate samples.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>(<span class="fu">library</span>(here))</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>(<span class="fu">library</span>(GenomicRanges))</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Read in peak calls</span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>peak_files <span class="ot">&lt;-</span> <span class="fu">list.files</span>(<span class="fu">here</span>(<span class="st">"data"</span>, <span class="st">"04_callpeak"</span>), <span class="at">pattern=</span><span class="st">"*.narrowPeak"</span>, <span class="at">full.names=</span><span class="cn">TRUE</span>)</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(peak_files) <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">".narrowPeak"</span>, <span class="st">""</span>, <span class="fu">basename</span>(peak_files))</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>peak_calls <span class="ot">&lt;-</span> <span class="fu">lapply</span>(peak_files, rtracklayer<span class="sc">::</span>import)</span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Separate into groups </span></span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a>control_calls <span class="ot">&lt;-</span> peak_calls[<span class="fu">c</span>(<span class="st">"control1"</span>, <span class="st">"control2"</span>, <span class="st">"control3"</span>)]</span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a>treatment_calls <span class="ot">&lt;-</span> peak_calls[<span class="fu">c</span>(<span class="st">"treatment1"</span>, <span class="st">"treatment2"</span>, <span class="st">"treatment3"</span>)]</span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a>control_calls <span class="ot">&lt;-</span> <span class="fu">GRangesList</span>(control_calls)</span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a>treatment_calls <span class="ot">&lt;-</span> <span class="fu">GRangesList</span>(treatment_calls)</span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute coverage across ranges</span></span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true" tabindex="-1"></a>control_coverage <span class="ot">&lt;-</span> <span class="fu">coverage</span>(control_calls)</span>
<span id="cb6-18"><a href="#cb6-18" aria-hidden="true" tabindex="-1"></a>treatment_coverage <span class="ot">&lt;-</span> <span class="fu">coverage</span>(treatment_calls)</span>
<span id="cb6-19"><a href="#cb6-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-20"><a href="#cb6-20" aria-hidden="true" tabindex="-1"></a><span class="co"># Determine regions with coverage in N replicates -- here require all 3 to have the same peaks</span></span>
<span id="cb6-21"><a href="#cb6-21" aria-hidden="true" tabindex="-1"></a>control_covered <span class="ot">&lt;-</span> <span class="fu">GRanges</span>(<span class="fu">slice</span>(control_coverage, <span class="at">lower=</span><span class="fu">length</span>(control_calls), <span class="at">rangesOnly=</span><span class="cn">TRUE</span>))</span>
<span id="cb6-22"><a href="#cb6-22" aria-hidden="true" tabindex="-1"></a>treatment_covered <span class="ot">&lt;-</span> <span class="fu">GRanges</span>(<span class="fu">slice</span>(treatment_coverage, <span class="at">lower=</span><span class="fu">length</span>(treatment_calls), <span class="at">rangesOnly=</span><span class="cn">TRUE</span>))</span>
<span id="cb6-23"><a href="#cb6-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-24"><a href="#cb6-24" aria-hidden="true" tabindex="-1"></a><span class="co"># Close gaps in the resulting regions -- min.gapwidth can be adjusted </span></span>
<span id="cb6-25"><a href="#cb6-25" aria-hidden="true" tabindex="-1"></a>control_consensus <span class="ot">&lt;-</span> <span class="fu">reduce</span>(control_covered, <span class="at">min.gapwidth=</span><span class="dv">101</span>)</span>
<span id="cb6-26"><a href="#cb6-26" aria-hidden="true" tabindex="-1"></a>treatment_consensus <span class="ot">&lt;-</span> <span class="fu">reduce</span>(treatment_covered, <span class="at">min.gapwidth=</span><span class="dv">101</span>)</span>
<span id="cb6-27"><a href="#cb6-27" aria-hidden="true" tabindex="-1"></a>all_consensus <span class="ot">&lt;-</span> <span class="fu">union</span>(control_consensus, treatment_consensus)</span>
<span id="cb6-28"><a href="#cb6-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-29"><a href="#cb6-29" aria-hidden="true" tabindex="-1"></a><span class="co"># Export as BED files</span></span>
<span id="cb6-30"><a href="#cb6-30" aria-hidden="true" tabindex="-1"></a>rtracklayer<span class="sc">::</span><span class="fu">export.bed</span>(control_consensus, <span class="fu">here</span>(<span class="st">"data"</span>, <span class="st">"04_callpeak"</span>, <span class="st">"control-consensus.bed"</span>))</span>
<span id="cb6-31"><a href="#cb6-31" aria-hidden="true" tabindex="-1"></a>rtracklayer<span class="sc">::</span><span class="fu">export.bed</span>(treatment_consensus, <span class="fu">here</span>(<span class="st">"data"</span>, <span class="st">"04_callpeak"</span>, <span class="st">"treatment-consensus.bed"</span>))</span>
<span id="cb6-32"><a href="#cb6-32" aria-hidden="true" tabindex="-1"></a>rtracklayer<span class="sc">::</span><span class="fu">export.bed</span>(all_consensus, <span class="fu">here</span>(<span class="st">"data"</span>, <span class="st">"04_callpeak"</span>, <span class="st">"consensus.bed"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="create-raw-signal-tracks" class="level2">
<h2 class="anchored" data-anchor-id="create-raw-signal-tracks">Create raw signal tracks</h2>
<p>Genrich outputs a bedgraph-ish file using the <code>-k</code> flag that can be used to quickly create a a bigwig file of the raw signal which can be viewed in IGV. I find it useful to look at the raw signal tracks after peak calling so I can assess the absolute magnitude of the pileup that went into calling a peak. Then, if the peaks look too permissive or too stringent, adjust the peak calls.</p>
<p>Creating bigwigs from the bedgraph-ish files can be accomplished using <code>awk</code> and <a href="https://www.encodeproject.org/software/bedgraphtobigwig/">bedGraphToBigWig</a> from UCSC.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co">#!/usr/bin/env bash</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert the pileups computed by Genrich from bedGraph-ish files to bigwigs</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="co"># ----------------------------------------------------------------------------</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a><span class="bu">set</span> <span class="at">-Eeou</span> pipefail</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a><span class="va">SAMPLES</span><span class="op">=</span>/path/to/sample-names.txt   </span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a><span class="va">PEAKS</span><span class="op">=</span>/path/to/03_callpeak</span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a><span class="va">OUT</span><span class="op">=</span>/path/to/04_bg2bw</span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a><span class="va">CHROM_SIZES</span><span class="op">=</span>/path/to/chrom.sizes</span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a><span class="va">JOBS</span><span class="op">=</span>6</span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> <span class="va">$OUT</span></span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-16"><a href="#cb7-16" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"Extracting bedGraphs from experimental signal..."</span></span>
<span id="cb7-17"><a href="#cb7-17" aria-hidden="true" tabindex="-1"></a><span class="ex">parallel</span> <span class="at">--jobs</span> <span class="va">$JOBS</span> <span class="dt">\</span></span>
<span id="cb7-18"><a href="#cb7-18" aria-hidden="true" tabindex="-1"></a>  <span class="st">"awk 'NR &gt; 2 { print </span><span class="dt">\$</span><span class="st">1, </span><span class="dt">\$</span><span class="st">2, </span><span class="dt">\$</span><span class="st">3, </span><span class="dt">\$</span><span class="st">4 }' </span><span class="va">$PEAKS</span><span class="st">/{}.pileup.txt | </span><span class="dt">\</span></span>
<span id="cb7-19"><a href="#cb7-19" aria-hidden="true" tabindex="-1"></a><span class="st">  LC_ALL=C sort -k1,1 -k2,2n &gt; </span><span class="va">$OUT</span><span class="st">/{}.sorted.bedGraph"</span> :::: <span class="va">$SAMPLES</span></span>
<span id="cb7-20"><a href="#cb7-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-21"><a href="#cb7-21" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"Creating bigwigs from pileups..."</span></span>
<span id="cb7-22"><a href="#cb7-22" aria-hidden="true" tabindex="-1"></a><span class="ex">parallel</span> <span class="at">--jobs</span> <span class="va">$JOBS</span> <span class="st">"bedGraphToBigWig </span><span class="va">$OUT</span><span class="st">/{}.sorted.bedGraph </span><span class="va">$CHROM_SIZES</span><span class="st"> </span><span class="va">$OUT</span><span class="st">/{}.bw"</span> :::: <span class="va">$SAMPLES</span></span>
<span id="cb7-23"><a href="#cb7-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-24"><a href="#cb7-24" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"Cleaning up intermediate bedGraphs..."</span></span>
<span id="cb7-25"><a href="#cb7-25" aria-hidden="true" tabindex="-1"></a><span class="fu">find</span> <span class="va">$OUT</span> <span class="at">-type</span> f <span class="at">-name</span> <span class="st">"*.bedGraph"</span> <span class="at">-delete</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="chrom-sizes-files" class="level3">
<h3 class="anchored" data-anchor-id="chrom-sizes-files">Chrom sizes files</h3>
<p>You can supply a URL to <code>bedGraphToBigWig</code> to get the chrom.sizes if you’re genome is supported. Otherwise, you can create a chrom.sizes file using another UCSC utility, <code>faSize</code>, and your genome fasta.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="ex">faSize</span> <span class="at">-detailed</span> <span class="at">-tab</span> genome.fasta <span class="op">&gt;</span> genome.chrom.sizes</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="atacseqqc" class="level2">
<h2 class="anchored" data-anchor-id="atacseqqc">ATACSeqQC</h2>
<p><strong>TODO: complete this section</strong></p>
<p>The <a href="https://bioconductor.org/packages/release/bioc/vignettes/ATACseqQC/inst/doc/ATACseqQC.html">ATACseqQC</a> R package can be used to calculate various QC stats on the filtered BAM files. These should be assessed closely and compared to the <a href="https://www.encodeproject.org/atac-seq/">ENCODE data standards</a>.</p>
</section>
<section id="differential-abundance-using-csaw-and-edger" class="level2">
<h2 class="anchored" data-anchor-id="differential-abundance-using-csaw-and-edger">Differential abundance using <code>csaw</code> and <code>edgeR</code></h2>
<p><strong>TODO: elaborate on this section</strong></p>
<p><a href="https://bioconductor.org/books/release/csawBook/">csaw</a> can be used to perform differential abundance over sliding windows <em>or</em> by counting reads in the consensus peaks defined above. This <a href="https://epigeneticsandchromatin.biomedcentral.com/articles/10.1186/s13072-020-00342-y">paper</a> contains some code for performing and assessing ATAC-seq DA analysis using both methods.</p>
<p>I usually generate read counts over the consensus peaks using <code>featureCounts</code> from the <a href="https://bioconductor.org/packages/devel/bioc/vignettes/Rsubread/inst/doc/SubreadUsersGuide.pdf">Rsubread</a> package. An SAF formatted file can be generated easily from the “consensus.bed” file. To get the read counts to align with the Genrich called peaks, set the <code>-read2Pos 5</code> flag to count alignments over the 5’ end of reads.</p>
</section>
<section id="create-tss-and-region-plots-with-deeptools" class="level2">
<h2 class="anchored" data-anchor-id="create-tss-and-region-plots-with-deeptools">Create TSS and region plots with <code>deeptools</code></h2>
<p><a href="https://deeptools.readthedocs.io/en/latest/">deeptools</a> can be used to compute normalized coverage files in bigwig format for visualization in IGV. Normalization in <code>deeptools</code> can be really context dependent and what normalization strategy to use can often be unclear.</p>
<p>If you’ve performed differential abundance analysis and have global scaling normalization factors, they can be used at this step to generate <a href="https://www.biostars.org/p/413626/#414440">TMM normalized</a> bigwigs. Below, scale each bigwig file using TMM scaling factors computed by <code>edgeR::normLibSizes()</code></p>
<p>First, compute the reciprocal of the TMM scaling factors in R:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># This assumes you have used featureCounts and imported into edgeR DGEList object 'y'</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>y <span class="ot">&lt;-</span> <span class="fu">normLibSizes</span>(y)</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute the scale factors for each library</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>sf <span class="ot">&lt;-</span> y<span class="sc">$</span>samples<span class="sc">$</span>norm.factors <span class="sc">*</span> y<span class="sc">$</span>samples<span class="sc">$</span>lib.size <span class="sc">/</span> <span class="fl">1e6</span>   </span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Take the inverse of the scale factors for use in deeptools</span></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>inv_sf <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">/</span> sf</span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Write out to a csv file -- assuming the colnames match the file basenames</span></span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a><span class="fu">fwrite</span>(</span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.frame</span>(<span class="at">sample_name =</span> <span class="fu">colnames</span>(y), </span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a>             <span class="at">scale_factor=</span>inv_sf), </span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a>  <span class="st">"scale-factors.csv"</span>, </span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a>  <span class="at">col.names=</span><span class="cn">FALSE</span></span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Then, to compute TMM scaled bigwigs using <code>deeptools</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co">#!/usr/bin/env bash</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute TMM normalized coverage bigwigs for visualization</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a><span class="co"># 'deeptools' mamba env must be activated before running</span></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a><span class="co"># -----------------------------------------------------------------------------</span></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a><span class="va">SAMPLES</span><span class="op">=</span>/path/to/sample-names.txt</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a><span class="va">SCALE_FACTORS</span><span class="op">=</span>/path/to/scale-factors.csv</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a><span class="va">BLACKLIST</span><span class="op">=</span>/path/to/excluderanges.bed</span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a><span class="va">CODING</span><span class="op">=</span>/path/to/coding.bed</span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a><span class="va">PEAKS</span><span class="op">=</span>/path/to/consensus.bed</span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a><span class="va">BAM</span><span class="op">=</span>/path/to/02_align</span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a><span class="va">OUT</span><span class="op">=</span>/path/to/05_deeptools</span>
<span id="cb10-14"><a href="#cb10-14" aria-hidden="true" tabindex="-1"></a><span class="va">JOBS</span><span class="op">=</span>6</span>
<span id="cb10-15"><a href="#cb10-15" aria-hidden="true" tabindex="-1"></a><span class="va">THREADS</span><span class="op">=</span>4</span>
<span id="cb10-16"><a href="#cb10-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-17"><a href="#cb10-17" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> <span class="va">$OUT</span></span>
<span id="cb10-18"><a href="#cb10-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-19"><a href="#cb10-19" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"Position sorting BAMs..."</span></span>
<span id="cb10-20"><a href="#cb10-20" aria-hidden="true" tabindex="-1"></a><span class="ex">parallel</span> <span class="at">--jobs</span> <span class="va">$JOBS</span> <span class="st">"samtools sort -@</span><span class="va">$THREADS</span><span class="st"> -m4G -o </span><span class="va">$OUT</span><span class="st">/{}.sorted.bam </span><span class="va">$BAM</span><span class="st">/{}.bam"</span> :::: <span class="va">$SAMPLES</span></span>
<span id="cb10-21"><a href="#cb10-21" aria-hidden="true" tabindex="-1"></a><span class="ex">parallel</span> <span class="at">--jobs</span> <span class="va">$JOBS</span> <span class="st">"samtools index </span><span class="va">$OUT</span><span class="st">/{}.sorted.bam"</span> :::: <span class="va">$SAMPLES</span></span>
<span id="cb10-22"><a href="#cb10-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-23"><a href="#cb10-23" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"Computing coverage over all BAM files..."</span> </span>
<span id="cb10-24"><a href="#cb10-24" aria-hidden="true" tabindex="-1"></a><span class="ex">parallel</span> <span class="at">--csv</span> <span class="at">--jobs</span> <span class="va">$JOBS</span> <span class="dt">\</span></span>
<span id="cb10-25"><a href="#cb10-25" aria-hidden="true" tabindex="-1"></a>    <span class="st">"bamCoverage --bam </span><span class="va">$BAM</span><span class="st">/{1}.sorted.bam </span><span class="dt">\</span></span>
<span id="cb10-26"><a href="#cb10-26" aria-hidden="true" tabindex="-1"></a><span class="st">                 --outFileName </span><span class="va">$OUT</span><span class="st">/{1}.bw </span><span class="dt">\</span></span>
<span id="cb10-27"><a href="#cb10-27" aria-hidden="true" tabindex="-1"></a><span class="st">                 --scaleFactor {2} </span><span class="dt">\</span></span>
<span id="cb10-28"><a href="#cb10-28" aria-hidden="true" tabindex="-1"></a><span class="st">                 --outFileFormat 'bigwig' </span><span class="dt">\</span></span>
<span id="cb10-29"><a href="#cb10-29" aria-hidden="true" tabindex="-1"></a><span class="st">                 --blackListFileName </span><span class="va">$BLACKLIST</span><span class="st"> </span><span class="dt">\</span></span>
<span id="cb10-30"><a href="#cb10-30" aria-hidden="true" tabindex="-1"></a><span class="st">                 --numberOfProcessors </span><span class="va">$THREADS</span><span class="st"> </span><span class="dt">\</span></span>
<span id="cb10-31"><a href="#cb10-31" aria-hidden="true" tabindex="-1"></a><span class="st">                 --ignoreDuplicates"</span> :::: <span class="va">$SCALE_FACTORS</span></span>
<span id="cb10-32"><a href="#cb10-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-33"><a href="#cb10-33" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"Computing matrix over TSS regions..."</span></span>
<span id="cb10-34"><a href="#cb10-34" aria-hidden="true" tabindex="-1"></a><span class="ex">computeMatrix</span> reference-point <span class="dt">\</span></span>
<span id="cb10-35"><a href="#cb10-35" aria-hidden="true" tabindex="-1"></a> <span class="at">--regionsFileName</span>  <span class="dt">\</span></span>
<span id="cb10-36"><a href="#cb10-36" aria-hidden="true" tabindex="-1"></a> <span class="at">--scoreFileName</span> <span class="va">$OUT</span>/<span class="pp">*</span>.bw <span class="dt">\</span></span>
<span id="cb10-37"><a href="#cb10-37" aria-hidden="true" tabindex="-1"></a> <span class="at">--outFileName</span> <span class="va">$OUT</span>/tss.mat.gz <span class="dt">\</span></span>
<span id="cb10-38"><a href="#cb10-38" aria-hidden="true" tabindex="-1"></a> <span class="at">--referencePoint</span> TSS <span class="dt">\</span></span>
<span id="cb10-39"><a href="#cb10-39" aria-hidden="true" tabindex="-1"></a> <span class="at">--beforeRegionStartLength</span> 3000 <span class="dt">\</span></span>
<span id="cb10-40"><a href="#cb10-40" aria-hidden="true" tabindex="-1"></a> <span class="at">--afterRegionStartLength</span> 3000 <span class="dt">\</span></span>
<span id="cb10-41"><a href="#cb10-41" aria-hidden="true" tabindex="-1"></a> <span class="at">--blackListFileName</span> <span class="va">$BLACKLIST</span> <span class="dt">\</span></span>
<span id="cb10-42"><a href="#cb10-42" aria-hidden="true" tabindex="-1"></a> <span class="at">--missingDataAsZero</span> <span class="dt">\</span></span>
<span id="cb10-43"><a href="#cb10-43" aria-hidden="true" tabindex="-1"></a> <span class="at">--numberOfProcessors</span> <span class="va">$JOBS</span></span>
<span id="cb10-44"><a href="#cb10-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-45"><a href="#cb10-45" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"Computing coverage over peak BEDs..."</span></span>
<span id="cb10-46"><a href="#cb10-46" aria-hidden="true" tabindex="-1"></a><span class="ex">computeMatrix</span> reference-point <span class="dt">\</span></span>
<span id="cb10-47"><a href="#cb10-47" aria-hidden="true" tabindex="-1"></a>  <span class="at">--regionsFileName</span> <span class="va">$PEAKS</span> <span class="dt">\</span></span>
<span id="cb10-48"><a href="#cb10-48" aria-hidden="true" tabindex="-1"></a>  <span class="at">--scoreFileName</span> <span class="va">$OUT</span>/<span class="pp">*</span>.bw <span class="dt">\</span></span>
<span id="cb10-49"><a href="#cb10-49" aria-hidden="true" tabindex="-1"></a>  <span class="at">--outFileName</span> <span class="va">$OUT</span>/peak.mat.gz <span class="dt">\</span></span>
<span id="cb10-50"><a href="#cb10-50" aria-hidden="true" tabindex="-1"></a>  <span class="at">--referencePoint</span> <span class="st">'center'</span> <span class="dt">\</span></span>
<span id="cb10-51"><a href="#cb10-51" aria-hidden="true" tabindex="-1"></a>  <span class="at">--beforeRegionStartLength</span> 3000 <span class="dt">\</span></span>
<span id="cb10-52"><a href="#cb10-52" aria-hidden="true" tabindex="-1"></a>  <span class="at">--afterRegionStartLength</span> 3000 <span class="dt">\</span></span>
<span id="cb10-53"><a href="#cb10-53" aria-hidden="true" tabindex="-1"></a>  <span class="at">--blackListFileName</span> <span class="va">$BLACKLIST</span> <span class="dt">\</span></span>
<span id="cb10-54"><a href="#cb10-54" aria-hidden="true" tabindex="-1"></a>  <span class="at">--missingDataAsZero</span> <span class="dt">\</span></span>
<span id="cb10-55"><a href="#cb10-55" aria-hidden="true" tabindex="-1"></a>  <span class="at">--numberOfProcessors</span> <span class="va">$JOBS</span></span>
<span id="cb10-56"><a href="#cb10-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-57"><a href="#cb10-57" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"Plotting heatmap of TSS enrichment..."</span></span>
<span id="cb10-58"><a href="#cb10-58" aria-hidden="true" tabindex="-1"></a><span class="ex">plotHeatmap</span> <span class="at">-m</span> <span class="va">$OUT</span>/tss.mat.gz <span class="dt">\</span></span>
<span id="cb10-59"><a href="#cb10-59" aria-hidden="true" tabindex="-1"></a>  <span class="at">-o</span> <span class="va">$OUT</span>/tss-heatmap.pdf <span class="dt">\</span></span>
<span id="cb10-60"><a href="#cb10-60" aria-hidden="true" tabindex="-1"></a>  <span class="at">--colorMap</span> <span class="st">"viridis"</span> <span class="dt">\</span></span>
<span id="cb10-61"><a href="#cb10-61" aria-hidden="true" tabindex="-1"></a>  <span class="at">--boxAroundHeatmaps</span> <span class="st">'no'</span> <span class="dt">\</span></span>
<span id="cb10-62"><a href="#cb10-62" aria-hidden="true" tabindex="-1"></a>  <span class="at">--samplesLabel</span> <span class="st">"Ctl 1"</span> <span class="st">"Ctl 2"</span> <span class="st">"Ctl 3"</span> <span class="st">"Trt 1"</span> <span class="st">"Trt 2"</span> <span class="st">"Trt 3"</span> <span class="dt">\</span></span>
<span id="cb10-63"><a href="#cb10-63" aria-hidden="true" tabindex="-1"></a>  <span class="at">--regionsLabel</span> <span class="st">"Protein Coding Genes"</span> <span class="dt">\</span></span>
<span id="cb10-64"><a href="#cb10-64" aria-hidden="true" tabindex="-1"></a>  <span class="at">--yAxisLabel</span> <span class="st">"RPGC"</span> <span class="dt">\</span></span>
<span id="cb10-65"><a href="#cb10-65" aria-hidden="true" tabindex="-1"></a>  <span class="at">--perGroup</span> <span class="dt">\</span></span>
<span id="cb10-66"><a href="#cb10-66" aria-hidden="true" tabindex="-1"></a>  <span class="at">--plotFileFormat</span> <span class="st">"pdf"</span> <span class="dt">\</span></span>
<span id="cb10-67"><a href="#cb10-67" aria-hidden="true" tabindex="-1"></a>  <span class="at">--dpi</span> 300 <span class="dt">\</span></span>
<span id="cb10-68"><a href="#cb10-68" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-69"><a href="#cb10-69" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"Plotting heatmap of MACS peak enrichment..."</span></span>
<span id="cb10-70"><a href="#cb10-70" aria-hidden="true" tabindex="-1"></a><span class="ex">plotHeatmap</span> <span class="at">-m</span> <span class="va">$OUT</span>/peak.mat.gz <span class="dt">\</span></span>
<span id="cb10-71"><a href="#cb10-71" aria-hidden="true" tabindex="-1"></a>  <span class="at">-o</span> <span class="va">$OUT</span>/peak-heatmap.pdf <span class="dt">\</span></span>
<span id="cb10-72"><a href="#cb10-72" aria-hidden="true" tabindex="-1"></a>  <span class="at">--colorMap</span> <span class="st">"viridis"</span> <span class="dt">\</span></span>
<span id="cb10-73"><a href="#cb10-73" aria-hidden="true" tabindex="-1"></a>  <span class="at">--boxAroundHeatmaps</span> <span class="st">'no'</span> <span class="dt">\</span></span>
<span id="cb10-74"><a href="#cb10-74" aria-hidden="true" tabindex="-1"></a>  <span class="at">--samplesLabel</span> <span class="st">"Ctl 1"</span> <span class="st">"Ctl 2"</span> <span class="st">"Ctl 3"</span> <span class="st">"Trt 1"</span> <span class="st">"Trt 2"</span> <span class="st">"Trt 3"</span> <span class="dt">\</span></span>
<span id="cb10-75"><a href="#cb10-75" aria-hidden="true" tabindex="-1"></a>  <span class="at">--regionsLabel</span> <span class="st">"Consensus Peak Calls"</span> <span class="dt">\</span></span>
<span id="cb10-76"><a href="#cb10-76" aria-hidden="true" tabindex="-1"></a>  <span class="at">--refPointLabel</span> <span class="st">"Center of Peak"</span> <span class="dt">\</span></span>
<span id="cb10-77"><a href="#cb10-77" aria-hidden="true" tabindex="-1"></a>  <span class="at">--yAxisLabel</span> <span class="st">"RPGC"</span> <span class="dt">\</span></span>
<span id="cb10-78"><a href="#cb10-78" aria-hidden="true" tabindex="-1"></a>  <span class="at">--perGroup</span> <span class="dt">\</span></span>
<span id="cb10-79"><a href="#cb10-79" aria-hidden="true" tabindex="-1"></a>  <span class="at">--plotFileFormat</span> <span class="st">"pdf"</span> <span class="dt">\</span></span>
<span id="cb10-80"><a href="#cb10-80" aria-hidden="true" tabindex="-1"></a>  <span class="at">--dpi</span> 300 <span class="dt">\</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="getting-tss-coding-region-ranges" class="level3">
<h3 class="anchored" data-anchor-id="getting-tss-coding-region-ranges">Getting TSS coding region ranges</h3>
<p>For ATAC-seq, I plot the enrichment over the TSS region for each sample as well as the coverage centered over the consensus peaks. To generate a BED file for protein coding gene regions (“coding.bed” above) in R:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Get mouse annotation GTF, for example</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>url <span class="ot">&lt;-</span> <span class="st">"https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M37/gencode.vM37.annotation.gtf.gz"</span></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>gtf <span class="ot">&lt;-</span> rtracklayer<span class="sc">::</span><span class="fu">import</span>(url)</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract the protein coding gene ranges only</span></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>coding <span class="ot">&lt;-</span> gtf[gtf<span class="sc">$</span>gene_type <span class="sc">==</span> <span class="st">"protein_coding"</span> <span class="sc">&amp;</span> gtf<span class="sc">$</span>type <span class="sc">==</span> <span class="st">"gene"</span>, ]</span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a><span class="co"># rtracklayer complains if score is NA or non-numeric</span></span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>coding<span class="sc">$</span>score <span class="ot">&lt;-</span> 1L</span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Export to a BED file for use in deeptools</span></span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a>rtracklayer<span class="sc">::</span><span class="fu">export.bed</span>(coding, <span class="fu">here</span>(<span class="st">"doc"</span>, <span class="st">"gencode.vM37.coding.bed"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="motif-analysis-with-meme" class="level2">
<h2 class="anchored" data-anchor-id="motif-analysis-with-meme">Motif analysis with <code>MEME</code></h2>
<p>The <a href="https://meme-suite.org/meme/">MEME suite</a> can be used to perform motif analysis on peaks that have been determined to have differential abundance. You can use R to generate fasta files for these regions based on GRanges extracted from DA analysis. These fasta files can then be used as input to <code>XSTREME</code>, which performs de novo motif discovery, enrichment analysis, and comparisons with known motif databases.</p>
<p>For example, you can extract fasta regions like so (this uses hg38 as an example):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>(<span class="fu">library</span>(data.table))</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>(<span class="fu">library</span>(Biostrings))</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>(<span class="fu">library</span>(BSgenome.Hsapiens.UCSC.hg38))</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>(<span class="fu">library</span>(GenomicRanges))</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Assuming you have a file from DA analysis that contains genomic coordinates</span></span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a><span class="co"># of peaks and annotations - for example from ChIPseeker::annotatePeak()</span></span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>peaks <span class="ot">&lt;-</span> <span class="fu">fread</span>(<span class="st">"annotated-peaks.tsv"</span>)</span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>peaks <span class="ot">&lt;-</span> <span class="fu">makeGRangesFromDataFrame</span>(peaks, <span class="at">keep.extra.columns =</span> <span class="cn">TRUE</span>)</span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract promoter peaks -- this can be any filtering you're interested in </span></span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a>promoter_peaks <span class="ot">&lt;-</span> peaks[peaks<span class="sc">$</span>annotation <span class="sc">%like%</span> <span class="st">"Promoter"</span>]</span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-15"><a href="#cb12-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract and write out the hg38 sequences as fasta records</span></span>
<span id="cb12-16"><a href="#cb12-16" aria-hidden="true" tabindex="-1"></a>pmtr_seqs <span class="ot">&lt;-</span> <span class="fu">getSeq</span>(BSgenome.Hsapiens.UCSC.hg38, promoter_peaks)</span>
<span id="cb12-17"><a href="#cb12-17" aria-hidden="true" tabindex="-1"></a><span class="fu">writeXStringSet</span>(pmtr_seqs, <span class="at">filepath =</span> <span class="st">"promoter-peaks.fasta"</span>, <span class="at">format =</span> <span class="st">"fasta"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>These fasta records can then be used in <code>XSTREME</code> analysis from the MEME suite. The meme formatted databases of known motifs can be downloaded from MEME’s <a href="https://meme-suite.org/meme/db/motifs">website</a></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co">#!/usr/bin/env bash</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Run XSTREME motif analysis on promoter peak sequences</span></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a><span class="co"># -------------------------------------------------------------------------------------------------</span></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a><span class="va">FA</span><span class="op">=</span>/path/to/promoter-peaks.fasta</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a><span class="va">DB</span><span class="op">=</span>/path/to/HUMAN/HOCOMOCOv11_core_HUMAN_mono_meme_format.meme</span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a><span class="va">OUT</span><span class="op">=</span>/path/to/xstreme</span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a><span class="ex">xstreme</span> <span class="at">--oc</span> <span class="va">$OUT</span> <span class="at">--p</span> <span class="va">$FA</span> <span class="at">--seed</span> 123 <span class="at">--m</span> <span class="va">$DB</span> <span class="at">--no-pgc</span> <span class="at">--dna</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./differential-methylation-analysis.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Differential Methylation Analysis</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./references.html" class="pagination-link">
        <span class="nav-page-text">References</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>